rewrite-java/src/main/java/org/openrewrite/java/cleanup/ChainStringBuilderAppendCalls.java

```diff
@@ -0,0 +1,126 @@
+        return "String concatenation within calls to `StringBuilder.append()` causes unnecessary memory allocation. Except for concatenations of String literals, which are joined together at compile time. Replaces inefficient concatenations with chained calls to `StringBuilder.append()`.";
+    }
+
+    @Override
+    protected @Nullable TreeVisitor<?, ExecutionContext> getSingleSourceApplicableTest() {
+        return new UsesMethod<>(STRING_BUILDER_APPEND);
+    }
+
+    @Override
+    protected TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new JavaIsoVisitor<ExecutionContext>() {
+            @Override
+            public J.MethodInvocation visitMethodInvocation(J.MethodInvocation method, ExecutionContext ctx) {
+                if (STRING_BUILDER_APPEND.matches(method)) {
+                    List<Expression> arguments = method.getArguments();
+                    if (arguments.size() != 1) {
+                        return method;
+                    }
+
+                    List<Expression> flattenExpressions = new ArrayList<>();
+                    boolean flattenable = flatAdditiveExpressions(arguments.get(0), flattenExpressions);
+                    if (!flattenable) {
+                        return method;
+                    }
+
+                    if (flattenExpressions.stream().allMatch(exp -> exp instanceof J.Literal)) {
+                        return method;
+                    }
+
+                    // group expressions
+                    List<Expression> groups = new ArrayList<>();
+                    List<Expression> group = new ArrayList<>();
+                    for (Expression exp : flattenExpressions) {
+                        if (exp instanceof J.Literal) {
+                            group.add(exp);
+                        } else {
+                            addToGroups(group, groups);
+                            groups.add(exp);
+                        }
+                    }
+                    addToGroups(group, groups);
+
+                    J.MethodInvocation chainedMethods = method.withArguments(singletonList(groups.get(0)));
+                    for (int i = 1; i < groups.size(); i++) {
+                        chainedMethods = chainedMethods.withSelect(chainedMethods)
+                            .withArguments(singletonList(groups.get(i)))
+                            .withPrefix(Space.EMPTY);
+                    }
+
+                    return chainedMethods;
+                }
+
+                return method;
+            }
+        };
+    }
+
+    /**
+     * Concat an additive expression in a group and add to groups
+     */
+    private static void addToGroups(List<Expression> group, List<Expression> groups) {
+        if (!group.isEmpty()) {
+            groups.add(ExpressionUtils.additiveExpression(group));
+            group.clear();
+        }
+    }
+
+    private static boolean flatAdditiveExpressions(Expression expression, List<Expression> expressionList) {
```

Comment:
I think here and some other places where you have `instanceof` checks on `Expression` objects you may also want to call `Expression#unwrap()` first to automatically deal with `J.Parentheses`.

But in this specific case, since you already have a recursive method, you could also just add an `if` for `J.Parentheses`.

Reply:
Thanks for letting me know the functionality of `Expression#unwrap()`. 

Do you mean to cover a double-parenthesis case like this?
```java
sb.append(("A" + op + "B"));
```

I am also concerned that unnecessary unwrapping could make some problems like 
```
(1 + 2) * 3 
```
unwrapped to
```
1 + 2 * 3
```

Reply:
Also after unwrapping the LST would still have the "correct shape" (i.e. the `*` as the top-level node). But you would still have to be careful to serialize the correct thing back to the source code (I don't know if the JavaPrinter automatically adds "missing" parentheses, but that is an interesting question!).

I haven't looked carefully enough at the code to determine whether unwrapping actually makes sense here or not. Just a thought that struck me.

Reply:
will have a separate PR to handle this case of double parentheses since it's a change for another recipe.